bpf/bpf_lxc.c

```
@@ -272,19 +272,21 @@ static __always_inline int ipv6_l3_from_lxc(struct __ctx_buff *ctx,
 		policy_mark_skip(ctx);
 
 #ifdef ENABLE_NODEPORT
-		/* See comment in handle_ipv4_from_lxc(). */
-		if (ct_state.node_port) {
-			ctx->tc_index |= TC_INDEX_F_SKIP_RECIRCULATION;
-			ep_tail_call(ctx, CILIUM_CALL_IPV6_NODEPORT_REVNAT);
-			return DROP_MISSED_TAIL_CALL;
-		}
 # ifdef ENABLE_DSR
 		if (ct_state.dsr) {
 			ret = xlate_dsr_v6(ctx, tuple, l4_off);
 			if (ret != 0)
 				return ret;
-		}
+		} else
 # endif /* ENABLE_DSR */
+		/* See comment in handle_ipv4_from_lxc(). */
+		if (ct_state.node_port) {
+			send_trace_notify(ctx, TRACE_TO_NETWORK, SECLABEL,
+					  *dst_id, 0, 0, ret, monitor);
```

Curious if you considered putting this logic inside the `tail_rev_nodeport_lb[46]()` functions instead? That seems like it will better handle errors.

One simple reason would be, if you follow this logic, and the tail call fails, then we emit a `TO_NETWORK` event followed by dropping the traffic. A higher layer view on the network would be surprising then, because Cilium would be reporting that it handed the traffic off to the network, but then it actually silently dropped the traffic. Cilium probably won't emit a drop notification in that case, so the packets will just disappear immediately after Cilium reports that the stack is responsible for them.

The above case is fairly rare as it would require a tail call to miss. However there are subtly different issues if anything fails inside the tail call - for instance, if host firewall or the revnat code denies the traffic then we will emit both a `TO_NETWORK` (here) and a `ACT_DROP` notification (in the tail call). Usually elsewhere in Cilium, we emit exactly one of these events so that it's clear in the hubble output or higher layers exactly what happened to the packet. We can't both forward and drop the packet.

---

^^ This is more of a "correctness over time" type thing, so if it looks like switching to this approach will take significantly more effort (eg because of a change in the way that complexity is handled by the code when we shift the code around) then I'm OK with merging as-is and then addressing this afterwards.

---

Good concerns. @gandro has considered moving the `send_trace_notify()` invocation into the `tail_rev_nodeport_lb[46]()`. However, to make the aggregation work, we would need to pass the `monitor` value. One way to do this is to use `skb->cb[]` which is currently not used by the `tail_rev_nodeport_lb[46]()`, so we have enough space to store the value.

To unblock the issue, I would propose to address this in an immediate follow-up PR.